File: 139 Word Break/word_break.cs
public class Solution {
    public bool WordBreak(string s, ISet<string> wordDict) {
        bool[] dp = new bool[s.Length + 1];
        dp[0] = true;
        for (int i = 1; i <= s.Length; i++) {
            for (int j = i - 1; j >= 0; j--) {
                if (dp[j] && wordDict.Contains(s.Substring(j, i - j))) {
                    dp[i] = true;
                    break;
                }
            }
        }
        return dp[s.Length];
    }
} // Author: XC

File: 139 Word Break/word_break.cpp
class Solution {
public:
    bool wordBreak(string s, unordered_set<string>& wordDict) {
        vector<int> dp(s.size() + 1, 0);
        dp[0] = 1;
        for (int i = 1; i <= s.size(); i++) {
            for (int j = i - 1; j >= 0; j--) {
                if (dp[j] == 1 && wordDict.find(s.substr(j, i - j)) != wordDict.end()) {
                    dp[i] = 1;
                }
            }
        }
        return dp.back() == 1;
    }
}; // Author: XC
